rts/motoko-rts/src/stabilization/deserialization.rs

```diff
@@ -0,0 +1,186 @@
+mod scan_stack;
+pub mod stable_memory_access;
+
+use crate::{
+    memory::Memory,
+    stabilization::deserialization::scan_stack::STACK_EMPTY,
+    types::{FwdPtr, Tag, Value, TAG_ARRAY, TAG_ARRAY_SLICE_MIN, TAG_FWD_PTR},
+    visitor::visit_pointer_fields,
+};
+
+use self::{scan_stack::ScanStack, stable_memory_access::StableMemoryAccess};
+
+use super::{
+    clear_stable_memory,
+    graph_copy::{limit::InstructionLimit, GraphCopy},
+    layout::{deserialize, StableValue},
+    moc_stabilization_instruction_limit,
+};
+
+pub struct Deserialization {
+    from_space: StableMemoryAccess,
+    scan_stack: ScanStack,
+    stable_start: u64,
+    stable_size: u64,
+    stable_root: Option<Value>,
+    limit: InstructionLimit,
+}
+
+/// Helper type to pass serialization context instead of closures.
+pub struct DeserializationContext<'a, M: Memory> {
+    pub deserialization: &'a mut Deserialization,
+    pub mem: &'a mut M,
+}
+
+impl<'a, M: Memory> DeserializationContext<'a, M> {
+    fn new(
+        deserialization: &'a mut Deserialization,
+        mem: &'a mut M,
+    ) -> DeserializationContext<'a, M> {
+        DeserializationContext {
+            deserialization,
+            mem,
+        }
+    }
+}
+
+/// Graph-copy-based deserialization.
+/// Usage:
+/// ```
+/// let deserialization = Deserialization::start(mem, stable_start, stable_size);
+/// while !deserialization.is_completed() {
+///     deserialization.copy_increment();
+/// }
+/// ```
+impl Deserialization {
+    /// Start the deserialization, followed by a series of copy increments.
+    pub fn start<M: Memory>(mem: &mut M, stable_start: u64, stable_size: u64) -> Deserialization {
+        let from_space = StableMemoryAccess::open(stable_start, stable_size);
+        let scan_stack = unsafe { ScanStack::new(mem) };
+        let limit = InstructionLimit::new(unsafe { moc_stabilization_instruction_limit() });
```

Comment:
Note to my self: This limit needs to be reset to the lower update limit when subsequent destabilization increments are needed after the actual upgrade.

Reply:
This is now fixed.